src/worker/hwdetect.rs

```diff
@@ -1,24 +1,44 @@
+use crate::common::parser::{format_parse_error, p_uint, NomResult};
+use anyhow::anyhow;
+
 use tako::common::resources::{CpuId, NumOfCpus, ResourceDescriptor};
 
-pub fn parse_range(line: &str) -> anyhow::Result<Vec<CpuId>> {
-    let line = line.trim();
-    let mut result = Vec::new();
-    for part in line.split(',') {
-        let part = part.trim();
-        if part.contains('-') {
-            let ids: Vec<&str> = part.split('-').collect();
-            if ids.len() != 2 {
-                anyhow::bail!("Invalid format of range");
-            }
-            for i in ids[0].parse()?..=ids[1].parse()? {
-                result.push(i);
-            }
-        } else {
-            let id: CpuId = part.parse()?;
-            result.push(id);
-        }
-    }
-    Ok(result)
+use nom::bytes::complete::tag;
+use nom::combinator::all_consuming;
+use nom::combinator::{map_res, opt};
+use nom::multi::separated_list1;
+use nom::sequence::{preceded, tuple};
+
+fn remove_whitespace(s: &str) -> String {
+    s.split_whitespace().collect()
+}
+
+fn vec_from_range(start: CpuId, count: NumOfCpus) -> Vec<CpuId> {
+    (start..(start + count)).collect()
+}
+
+fn p_cpu_range(input: &str) -> NomResult<Vec<CpuId>> {
+    map_res(
+        tuple((p_uint, opt(preceded(tag("-"), p_uint)))),
+        |r| match r {
+            (u, None) => Ok(vec_from_range(u, 1)),
+            (u, Some(v)) if v >= u => Ok(vec_from_range(u, v - u + 1)),
+            _ => Err(anyhow!("Invalid format of range")),
+        },
+    )(input)
+}
+
+fn p_cpu_ranges(input: &str) -> NomResult<Vec<CpuId>> {
+    separated_list1(tag(","), p_cpu_range)(input)
+        .map(|(a, b)| (a, b.into_iter().flatten().collect()))
+}
+
+pub fn parse_range(input: &str) -> anyhow::Result<Vec<CpuId>> {
+    let f_input = remove_whitespace(input);
+    let res = all_consuming(p_cpu_ranges)(&*f_input)
```

Comment:
`f_input.as_str()`
and you can also remove the `let res` and just return the expression without `;`